File: proofload-java/src/main/java/io/github/matthewjones372/proofload/java/Goals.java
package io.github.matthewjones372.proofload.java;

import io.github.matthewjones372.proofload.Clock;
import io.github.matthewjones372.proofload.Goal;
import io.github.matthewjones372.proofload.StepName;
import java.time.Duration;

/**
 * What a run is asked to achieve, so the report says which goal missed and by
 * how much rather than a test failing on the first assertion that did.
 *
 * Kotlin writes these as {@code p99(placeOrder) under 200.milliseconds}. The
 * infix form has no Java spelling, and every call that builds one takes a
 * {@code StepName} or a {@code Share}, whose names mangle — so these are the
 * Java sources over {@link Judgements}.
 *
 * The clock defaults to {@link Clock#ResponseTime}: a goal written against
 * service time can be met by a generator that never sent the load.
 */
public final class Goals {

    private Goals() {
    }

    public static Goal p50Under(StepName step, Duration limit) {
        return p50Under(step, limit, Clock.ResponseTime);
    }

    public static Goal p50Under(StepName step, Duration limit, Clock clock) {
        return Judgements.percentileUnder(step.getName(), "p50", limit, clock);
    }

    public static Goal p95Under(StepName step, Duration limit) {
        return p95Under(step, limit, Clock.ResponseTime);
    }

    public static Goal p95Under(StepName step, Duration limit, Clock clock) {
        return Judgements.percentileUnder(step.getName(), "p95", limit, clock);
    }

    public static Goal p99Under(StepName step, Duration limit) {
        return p99Under(step, limit, Clock.ResponseTime);
    }

    public static Goal p99Under(StepName step, Duration limit, Clock clock) {
        return Judgements.percentileUnder(step.getName(), "p99", limit, clock);
    }

    public static Goal p999Under(StepName step, Duration limit) {
        return p999Under(step, limit, Clock.ResponseTime);
    }

    public static Goal p999Under(StepName step, Duration limit, Clock clock) {
        return Judgements.percentileUnder(step.getName(), "p999", limit, clock);
    }

    /** The share of requests that may fail, as a percentage: {@code 0.1} is a tenth of one percent. */
    public static Goal failureRateUnder(StepName step, double percent) {
        return Judgements.failureRateUnder(step.getName(), percent);
    }

    /** The same, asked of the whole run rather than one step. */
    public static Goal failureRateUnder(double percent) {
        return Judgements.failureRateUnder(null, percent);
    }

    /** The share of requests that were both fast enough and successful. */
    public static Goal goodputAtLeast(StepName step, Duration under, double percent) {
        return goodputAtLeast(step, under, percent, Clock.ResponseTime);
    }

    public static Goal goodputAtLeast(StepName step, Duration under, double percent, Clock clock) {
        return Judgements.goodputAtLeast(step.getName(), under, percent, clock);
    }
}

File: proofload-scala/src/main/scala/io/github/matthewjones372/proofload/scala/Goals.scala
package io.github.matthewjones372.proofload.scala

import io.github.matthewjones372.proofload.Clock
import io.github.matthewjones372.proofload.Goal
import io.github.matthewjones372.proofload.Share
import io.github.matthewjones372.proofload.StepName
import io.github.matthewjones372.proofload.java.Goals
import io.github.matthewjones372.proofload.java.Shares
import java.time.Duration as JavaDuration
import _root_.scala.concurrent.duration.FiniteDuration

/**
 * What a run is asked to achieve, so the report says which goal missed and by
 * how much rather than a test failing on the first assertion that did.
 *
 * Kotlin writes `p99(placeOrder) under 200.milliseconds`. The infix form has no
 * Java spelling, so `Goals` is a row of statics there; Scala has infix
 * extensions, and reading like the Kotlin is the whole point of these.
 *
 * The clock starts at response time, as it does in both other languages: a goal
 * written against service time can be met by a generator that never sent the
 * load.
 */
final class PercentileOf private[scala] (step: StepName, clock: Clock, goal: (StepName, JavaDuration, Clock) => Goal):

  /** The limit the percentile has to stay under. */
  infix def under(limit: FiniteDuration): Goal = goal(step, asJava(limit), clock)

  /** The same, for a caller already holding a `java.time.Duration`, which is what a ZIO one holds. */
  infix def under(limit: JavaDuration): Goal = goal(step, limit, clock)

/** How much of the run may fail, waiting for the share it has to stay under. */
final class FailuresOf private[scala] (goal: Double => Goal):

  infix def under(share: Share): Goal = goal(share.getPercent)

/** A step's good requests, waiting for the share of them a run has to reach. */
final class GoodputOf private[scala] (step: StepName, under: JavaDuration, clock: Clock):

  infix def atLeast(share: Share): Goal = Goals.goodputAtLeast(step, under, share.getPercent, clock)

def p50(step: StepName, of: Clock = Clock.ResponseTime): PercentileOf = PercentileOf(step, of, Goals.p50Under(_, _, _))

def p95(step: StepName, of: Clock = Clock.ResponseTime): PercentileOf = PercentileOf(step, of, Goals.p95Under(_, _, _))

def p99(step: StepName, of: Clock = Clock.ResponseTime): PercentileOf = PercentileOf(step, of, Goals.p99Under(_, _, _))

/** The tail, which a step under a thousand samples misses for want of having measured it. */
def p999(step: StepName, of: Clock = Clock.ResponseTime): PercentileOf =
  PercentileOf(step, of, Goals.p999Under(_, _, _))

/** The share of the whole run that may fail. */
def failureRate: FailuresOf = FailuresOf(percent => Goals.failureRateUnder(percent))

/** The same, asked of one step rather than of the run. */
def failureRate(step: StepName): FailuresOf = FailuresOf(percent => Goals.failureRateUnder(step, percent))

/** The share of a step's requests that were both fast enough and successful. */
def goodput(step: StepName, under: FiniteDuration, of: Clock = Clock.ResponseTime): GoodputOf =
  GoodputOf(step, asJava(under), of)

/**
 * The same, in the duration a ZIO caller already holds. Written as two arities
 * rather than one with a default, because Scala allows only one overloaded
 * alternative to declare defaults and the `FiniteDuration` one has them.
 */
def goodput(step: StepName, under: JavaDuration): GoodputOf = GoodputOf(step, under, Clock.ResponseTime)

def goodput(step: StepName, under: JavaDuration, of: Clock): GoodputOf = GoodputOf(step, under, of)

extension (share: Int)

  /** A share as a percentage: `1.percent` is one in a hundred. */
  def percent: Share = Shares.percent(share.toDouble)

extension (share: Double)

  def percent: Share = Shares.percent(share)
